consensus/src/quorum_store/batch_generator.rs

```diff
@@ -124,54 +211,80 @@ impl BatchGenerator {
             // Quorum store metrics
             counters::CREATED_EMPTY_BATCHES_COUNT.inc();
 
-            let duration = self.last_end_batch_time.elapsed().as_secs_f64();
             counters::EMPTY_BATCH_CREATION_DURATION
-                .observe_duration(Duration::from_secs_f64(duration));
-
+                .observe_duration(self.last_end_batch_time.elapsed());
             self.last_end_batch_time = Instant::now();
-            return None;
+            return vec![];
         } else {
             counters::PULLED_TXNS_COUNT.inc();
             counters::PULLED_TXNS_NUM.observe(pulled_txns.len() as f64);
         }
+        counters::BATCH_CREATION_DURATION.observe_duration(self.last_end_batch_time.elapsed());
 
-        // Quorum store metrics
-        counters::CREATED_BATCHES_COUNT.inc();
-
-        let duration = self.last_end_batch_time.elapsed().as_secs_f64();
-        counters::BATCH_CREATION_DURATION.observe_duration(Duration::from_secs_f64(duration));
-
-        counters::NUM_TXN_PER_BATCH.observe(pulled_txns.len() as f64);
-
-        let batch_id = self.batch_id;
-        self.batch_id.increment();
-        self.db
-            .save_batch_id(self.epoch, self.batch_id)
-            .expect("Could not save to db");
+        // Compute bucketed batches
+        let bucket_compute_start = Instant::now();
 
+        let exclude_txns_with_gas: Vec<_> = self
+            .batches_in_progress
+            .values()
+            .flatten()
+            .cloned()
+            .collect();
+        let account_min_gas = Self::build_account_min_gas(&exclude_txns_with_gas, &pulled_txns);
         let expiry_time = aptos_infallible::duration_since_epoch().as_micros() as u64
             + self.config.batch_expiry_gap_when_init_usecs;
-        let txn_summaries: Vec<_> = pulled_txns
-            .iter()
-            .map(|txn| TransactionSummary {
-                sender: txn.sender(),
-                sequence_number: txn.sequence_number(),
-            })
-            .collect();
 
-        let batch = Batch::new(
-            batch_id,
-            pulled_txns,
-            self.epoch,
-            expiry_time,
-            self.my_peer_id,
-        );
+        pulled_txns.sort_by_cached_key(|txn| {
```

Comment:
I am not sure implementation guarantees elements with the same values will not be reordered.